List Alternator clients in system.clients virtual table

nyh opened the pull request:
Before this series, the "system.clients" virtual table lists active connections (and their various properties, like client address, logged in username and client version) only for CQL requests. This series adds also Alternator clients to system.clients. One of the interesting use cases of this new feature is understanding exactly which SDK a user is using -without inspecting their application code.  Different SDKs pass different "User-Agent" headers in requests, and that User-Agent will be visible in the system.clients entries for Alternator requests as the "driver_name" field.

Unlike CQL where logged in username, driver name, etc. applies to a complete connection, in the Alternator API, different requests can theoretically be signed by different users and carry different headers but still arrive over the same HTTP connection. So instead of listing the currently open Alternator *connections*, we will list the currently active *requests*.

The first two patches introduce utilities that will be useful in the implementation. The third patch is the implementation itself (which is quite simple with the utility introduced in the second patch), and the fourth patch a regression test for the new feature.

Fixes #24993

This patch adds a new feature, so doesn't require a backport. Nevertheless, if we want it to get to existing customers more quickly to allow us to better understand their use case by reading the system.clients table, we may want to consider backporting this patch to existing branches. There is some risk involved in this patch, because it adds code that gets run on every Alternator request, so a bug on it can cause problems for every Alternator request.

nyh commented on alternator/scoped_item_list.hh:
```diff
@@ -0,0 +1,161 @@
+        return seastar::do_until([ gi ] { return gi->iter == gi->end; },
+            [ gi, fn = std::move(fn) ] {
+                fn((gi->iter++)->get());
+                return make_ready_future<>();
```
This comment is wrong (sorry, bot...) - one of the important features of seastar::do_until is that it it checks need_preempt() between iterations, so it's fine that the iterated function almost always (or in this case, always!) returns a ready future, and it still won't stall.

Because this code *unconditionally* returns a make_ready_future, it's even possible that the optimizer will recognize it is always ready and eliminate in the do_until template even the test for that - and leave only the preemption checks. If anyone is wondering why compiling Scylla is so slow - it is exactly because of these things :-) But it also results in very fast code.

In any case, calling maybe_yield here is unnecessary here, and will only make this code even slower (because even if maybe_yield decides not to yield, do_until will check again).

nyh commented:
> > So instead of listing the currently open Alternator connections, we will list the currently active requests
> 
> @nyh My 2 cents: as you noted, in CQL we list open connections and not the connections that are currently executing a query. It may be that most/all CQL connections are idle and they'd still be listed, whereas Alternator's connections are ephemeral (from the alternator_system.client's POV) and don't present the whole picture, because they skip inactive connections. This may be counter-intuitive to users used to reading system.clients, as they may expect to see everyone who is connected, and so perhaps this table should be renamed to alternator_system.active_requests or should just list open connections or maybe both cases are worth implementing.

Let me begin by repeating the **motivation** for this patch, which I also tried to explain in the cover letter. It's not really about being "compatible" with any documented meaning of system.client (the documentation there isn't great - https://github.com/scylladb/scylladb/issues/23264). Rather, we want to be able to connect to a live cluster and learn a bit about the type and identity of clients that use it. It was never about having a "complete list" of connections - it's always a snapshot in time, in any case. It's not even having an accurate graph of how many connections we have open at each time (we have a metric that counts the number of open http connections for that). It was only about getting a **sample** of clients, and see what kind of drivers they use, their IP address, and so on. So for the purpose of getting a sample of requests, it doesn't really matter how many requests we capture and whether or not we also capture idle connections - as long as the cluster isn't completely idle, we'll see some requests happening and get the driver information we're after.

Then there are the technical reason why it was more "natural" for me to implement it this way (listing ongoing requests, intead of connections). Two reasons:
1. In HTTP, requests are independent, so **some** of the system.clients fields like driver_version and username are, at least in theory (but usually not in practice) different for each request, so it's not obvious what you'll want to show for an idle connection not currently working on a request.
2. Alternator uses the Seastar HTTP server which handles the http and/or https connection, and only passes to Alternator the actual requests. So we can't add code to **Alternator** that does things on connection creation or destruction, and we'll need to modify Seastar if we wanted to do it.

> @nyh But is the above really the case? In the test I can see that the table where these new clients would be listed is named:
>
>internal_prefix + 'system.clients`
>
> So I assume that this table's name is different from system.clients and system.clients will not display both CQL and Alternator users simultaneously?

Yes, it is really the same `system.clients` table, and you would see there both Alternator and CQL connections (I didn't check it in this test, though). The thing is, this test is using the Alternator API, so to read a **system table** like "system.clients", which isn't a real Alternator table, the syntax we invented was adding that "internal prefix", `.scylla.alternator.`, and with this trick you can read a system table. This is documented in docs/alternator/new-apis.md in the "Accessing system tables from Scylla" section.

nyh commented:
In new version:
1. Improve comment in new class scoped_item_list, giving an example how to do it.
2. Fifth patch adds documentation on the new system.clients support in Alternator.
3. Rebased to current master.
 
I don't know why the CI tests failed, looks like something wrong with the infrastructure, but as usual I can't figure it from the tens of thousands of log lines. Let's hope this one will pass...

ptrsmrn commented:
> > > So instead of listing the currently open Alternator connections, we will list the currently active requests
> > 
> > 
> > @nyh My 2 cents: as you noted, in CQL we list open connections and not the connections that are currently executing a query. It may be that most/all CQL connections are idle and they'd still be listed, whereas Alternator's connections are ephemeral (from the alternator_system.client's POV) and don't present the whole picture, because they skip inactive connections. This may be counter-intuitive to users used to reading system.clients, as they may expect to see everyone who is connected, and so perhaps this table should be renamed to alternator_system.active_requests or should just list open connections or maybe both cases are worth implementing.
> 
> Let me begin by repeating the **motivation** for this patch, which I also tried to explain in the cover letter. It's not really about being "compatible" with any documented meaning of system.client (the documentation there isn't great - #23264). Rather, we want to be able to connect to a live cluster and learn a bit about the type and identity of clients that use it. It was never about having a "complete list" of connections - it's always a snapshot in time, in any case. It's not even having an accurate graph of how many connections we have open at each time (we have a metric that counts the number of open http connections for that). It was only about getting a **sample** of clients, and see what kind of drivers they use, their IP address, and so on. So for the purpose of getting a sample of requests, it doesn't really matter how many requests we capture and whether or not we also capture idle connections - as long as the cluster isn't completely idle, we'll see some requests happening and get the driver information we're after.
> 
> Then there are the technical reason why it was more "natural" for me to implement it this way (listing ongoing requests, intead of connections). Two reasons:
> 
> 1. In HTTP, requests are independent, so **some** of the system.clients fields like driver_version and username are, at least in theory (but usually not in practice) different for each request, so it's not obvious what you'll want to show for an idle connection not currently working on a request.
> 2. Alternator uses the Seastar HTTP server which handles the http and/or https connection, and only passes to Alternator the actual requests. So we can't add code to **Alternator** that does things on connection creation or destruction, and we'll need to modify Seastar if we wanted to do it.

Sure, thanks. I was just under an impression it could be more beneficial to save such an info into a table named a bit differently, so it doesn't confuse CQL developers/users/support, which are used to seeing "stable" connections - I think they use it to debug what may be wrong with the cluster and potentially which client may cause issues. But for sure it's better to have something than anything

> > @nyh But is the above really the case? In the test I can see that the table where these new clients would be listed is named:
> > internal_prefix + 'system.clients`
> > So I assume that this table's name is different from system.clients and system.clients will not display both CQL and Alternator users simultaneously?
> 
> Yes, it is really the same `system.clients` table, and you would see there both Alternator and CQL connections (I didn't check it in this test, though). The thing is, this test is using the Alternator API, so to read a **system table** like "system.clients", which isn't a real Alternator table, the syntax we invented was adding that "internal prefix", `.scylla.alternator.`, and with this trick you can read a system table. This is documented in docs/alternator/new-apis.md in the "Accessing system tables from Scylla" section.

Got it, thanks. So both CQL's and Alternator's users are going to be displayed there simultaneously. I wonder if it can make the table too big (having too many entries), as I'm afraid there was an issue opened against system.clients being slow (but I can't find it, maybe I made it up). Also, I don't think both CQL and Alternator will be enabled and both be heavily used on a single cluster, it's rather this or that I think.

mykaul commented:
@ptrsmrn - ref https://github.com/scylladb/scylla-enterprise/issues/5335 ?

nyh commented:
> @ptrsmrn - ref [scylladb/scylla-enterprise#5335](https://github.com/scylladb/scylla-enterprise/issues/5335) ?

Wow. There should be an issue about this here.

mykaul commented:
> system.clients is not a real table - it's a virtual table which is generated while read. The list of ongoing requests isn't even stored in memory all the time - it's a "boost intrusive" list, meaning every ongoing request has a few bytes of memory on its stack, and those few bytes are linked to each other, and when a scan of system.client comes in reads this list. Actually, the current implementation is slightly worse - it reads this entire list and copies it to into chunked_vector. But this will never be huge - I think it will be negligable compared to the real allocations some by all those ongoing requests (for Alternator, or idle connections for CQL).
> 
> In any case, all of this is using existing machinary that CQL uses to produce its system.clients virtual table, so I don't see any reason to fear why it will be worse in than CQL's system.clients.

We do have deployments with 2M(!) connections per node.

avikivity commented on utils/on_internal_error.cc:
```diff
@@ -19,4 +19,8 @@ namespace utils {
     seastar::on_internal_error(on_internal_error_logger, reason);
 }
 
+[[noreturn]] void on_fatal_internal_error(std::string_view reason) noexcept{
```
Bot is absolutely correct here.

avikivity commented on alternator/scoped_item_list.hh:
```diff
@@ -0,0 +1,181 @@
+#include "utils/on_internal_error.hh"
+
+template <typename T>
+class scoped_item_list {
```
There's nothing alternator specific here, it should be in utils/ and come with its own test.

avikivity commented on alternator/scoped_item_list.hh:
```diff
@@ -0,0 +1,181 @@
+    future<> for_each_gently(noncopyable_function<void(T&)> fn) {
+        _gentle_iterators.emplace_front(*this);
+        typename std::list<gentle_iterator>::iterator gi = _gentle_iterators.begin();
+        return seastar::do_until([ gi ] { return gi->iter == gi->end; },
```
I think it's okay to write this as a coroutine, accepting that the list is intended to be large, and for_each_gently is intended to be slow.

nyh replied:
I copied this for_each_gently code verbatim from generic_server.cc. Although I do understand what it does, I didn't want to rewrite it without reason, given it's already short and the do_until() is not what makes it complicated, why change it?

nyh commented on alternator/scoped_item_list.hh:
```diff
@@ -0,0 +1,181 @@
+ *
+ * When adding an item to the list with the emplace() method, a handle is
+ * returned which can be used to continue accessing the item. When this handle
+ * goes out of scope, the item is automatically removed from the list.
```
It's built on a boost::intrusive_list, indeed but has additional features described in the big comment in front of the class. One of them is the fact that the item is removed from the intrusive list when destructed. But the big feature is the ability to walk the list in preemptive code, with the risk that the walker prempts during the iteration.

Note that generic_server.cc had exactly the same capabilities, and much of the techniques I used in this patch are copied from it.

Perhaps the name scoped_item_list doesn't fully describe what this class is about (it certainly doesn't say anything about stall-free iteration). Maybe you have a better idea for a name?

avikivity commented on alternator/scoped_item_list.hh:
```diff
@@ -0,0 +1,181 @@
+
+    struct gentle_iterator {
+        list_type::iterator iter, end;
+        gentle_iterator(scoped_item_list<T>& s) : iter(s._list.begin()), end(s._list.end()) {}
```
Isn't it more accurate to sample end() each time (and capture s._list, not s.list.end())?

It ends up the same, because the iterator points at the list, but seems safer.

nyh replied:
Done (deleted from generic_server.hh). But I did NOT address your original comment yet - which was about the end() thing in the original code which I copied.

nyh commented:
CI in debug mode (with ASAN) caught a use-after-return in the reimplementation of  generic_server::server::for_each_gently in my latest version of the patch (which made the existing generic_server use the new scoped_item_list class). I'll need to debug it now.

nyh commented:
Wow, I let vscode's copilot automatically write for me some seemingly-unimportant boilerplate line of code, and it used the wrong capture (capture by reference instead of move) and led to this use-after-return bug. I guess AI isn't always I...

If you're curious, the broken code:
```c++
future<> server::for_each_gently(noncopyable_function<void(connection&)> fn) {
    return _connections_list.for_each_gently([&fn](std::reference_wrapper<connection> c) {
        fn(c.get());
    });
}
```
Note the capture `&fn` :-(

nyh commented:
CI pages say "Aborted by user @yaronkaikov". Why?
I won't run again - I'll wait until I implement the last missing piece (a unit test for the new C++ class) and then push a new version to get the CI to run.

nyh commented:
Pushed a version with an added unit test for the new `utils::scoped_item_list` class introduced in this series. Also rebased to resolve conflicts.

nyh commented:
ping, please review.

@avikivity I have done the big thing for which you asked in your review - namely that the new "scoped_item_list" class that I extracted from the CQL implementation to use in Alternator - is now also used in CQL instead of keeping the original duplicated code, and I also added unit tests for the new class.
